src/utils/io/progress_log.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use humantime::format_duration;
+use std::time::{Duration, Instant};
+
+use tracing::info;
+
+#[derive(Debug)]
+#[derive(Clone)]
+pub struct ProgressLog {
+    frequency: Duration,
+    message: String,
+    start: Instant,
+    last_logged: Instant,
+    size: u64,
+    total_size: u64,
+}
+
+impl ProgressLog {
+    pub fn builder() -> ProgressLogBuilder {
+        ProgressLogBuilder {
+            frequency: None,
+            message: None,
+        }
+    }
+
+    pub fn set(&mut self, current_size: u64) {
+        self.update(current_size);
+    }
+
+    pub fn inc(&mut self, value: u64) {
+        self.update(self.size + value);
+    }
+
+    fn update(&mut self, current_size: u64) {
+        self.size = current_size;
+        let now = Instant::now();
+        if (now - self.last_logged) > self.frequency {
+            let elapsed_secs = (now - self.start).as_secs_f64();
+            let elapsed_duration = format_duration(Duration::from_secs(elapsed_secs as u64));
+
+            let throughput = self.size as f64 / elapsed_secs;
+            let eta_secs = (self.total_size - self.size) as f64 / throughput;
+            let eta_duration = format_duration(Duration::from_secs(eta_secs as u64));
+
+            info!(
+                "{} {} (elapsed: {}, eta: {})",
+                self.message, current_size, elapsed_duration, eta_duration
+            );
+            self.last_logged = now;
+        }
+    }
+
+    fn finish(&mut self) {}
+
+    pub fn wrap_stream<S: futures_core::Stream>(&self, stream: S) -> ProgressLogIter<S> {
+        ProgressLogIter {
+            it: stream,
+            progress: self.clone(),
+        }
+    }
+}
+
+pub struct ProgressLogBuilder {
+    frequency: Option<Duration>,
+    message: Option<String>,
+}
```

Comment:
Do we _need_ a builder? They're often used for complex configuration of private fields, but our only consumer is internal, and we don't have that much config :shrug: